### PR TITLE
Adding more CODEOWNERS

### PR DESCRIPTION
**Pull request by TGWDB**
This is a first draft of adding some more CODEOWNERS in different places.
The underlying goal here is to ensure that Diffblue has two (or more)
in house CODEOWNERS for all regularly updated parts of the code. This
is to enable Diffblue to respond to the vast majority of the code base
and also reduce reliance upon a single person (within or without of
Diffblue).

**NOTE:** this is put up now for discussion and refinement. Consider this a provisional starting point.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
To all those being added as code owners, and for each directory: do you feel confident in deciding whether a change is the right one to make? That is to say: only in parts should others decide whether to approve of someone's code ownership. The main approval is to be provided by the person being added.

File: CODEOWNERS
Comment (tautschnig) [inline CODEOWNERS:57]: While at it, please remove my name from this one. I do not feel sufficiently competent/confident for SMT2.

**Review by chrisr-diffblue — APPROVED** [review]
Seems reasonable to me.

File: CODEOWNERS
Comment (TGWDB) [reply CODEOWNERS:57]: Done in pending commit.

**Comment by TGWDB** [comment]
> To all those being added as code owners, and for each directory: do you feel confident in deciding whether a change is the right one to make? That is to say: only in parts should others decide whether to approve of someone's code ownership. The main approval is to be provided by the person being added.

This is a great point and part of why I marked this RFC, I'd very much appreciate comment/approval/discussion from those effected (@peterschrammel @thomasspriggs @NlightNFotis @allredj @romainbrenguier ).

File: CODEOWNERS
Comment (allredj) [reply CODEOWNERS:57]: OK for me. I'm not active on SMT2 at the moment.

File: CODEOWNERS
Comment (thomasspriggs) [inline CODEOWNERS]: I can't say that I am any kind of authority on our assembly support, having never worked on its implementation in cbmc. However I have enough familiarity with cbmc front-ends in general, that I am not going to strenuously object to being made a codeowner of this, if we need more codeowners.

File: CODEOWNERS
Comment (thomasspriggs) [inline CODEOWNERS]: :+1:

File: CODEOWNERS
Comment (tautschnig) [reply CODEOWNERS]: Dare I question whether the above changes are in line with "regularly updated parts of the code?"

Comment (thomasspriggs) [reply CODEOWNERS]: This assembler sub folder has not been updated particularly regularly, as I read the commit history. Therefore adding more code owners for it is not in line with this PR's "regularly updated parts of the code" description. @TGWDB What is your view on this?

Comment (NlightNFotis) [reply CODEOWNERS]: I don't know anything about what's in `src/assembler` myself, so I don't think I'm particularly suited for being a code owner for this.

File: CODEOWNERS
Comment (NlightNFotis) [reply CODEOWNERS]: This, I've worked on a bit in the past, however given that the aim is sunsetting it in favour of the new SMT backend, I'm not sure what value code-ownership of this brings (I expect this to be a low traffic area of the code, but then again I might be mistaken).

File: CODEOWNERS
Comment (TGWDB) [reply CODEOWNERS]: I'll reduce some of these (commit ~~to come shortly~~ pushed). I agree they may not be "regular" as the PR suggests, on the other hand we had a problem with getting a PR approved by codeowner around the `ansi-c` part, so I'll add a little more here.

File: CODEOWNERS
Comment (TGWDB) [reply CODEOWNERS]: I'm tempted to keep this since we've had a few fixes and updates here, and with work on the new SMT side I can see some overlap in the future (bug fixes, minor refactors, etc.).

**Review by peterschrammel — APPROVED** [review]

**Review by kroening — CHANGES_REQUESTED** [review]
I am unhappy with the additions to /util and /goto-programs -- these are very sensitive, and being slow here is good.

File: CODEOWNERS
Comment (tautschnig) [reply CODEOWNERS]: If it helps, my role has very recently changed so that I now have a lot more time to work on CBMC.

**Comment by TGWDB** [comment]
> I am unhappy with the additions to /util and /goto-programs -- these are very sensitive, and being slow here is good.

I would like to clarify what would be acceptable here for you (and others)? To clarify, the goal of this PR is NOT to make changes fast, particularly to sensitive areas. The goal is to have redundancy so that we do not rely on 1 person (or less) that Diffblue can reach reliably for review/approval. There appear to be three obvious possible ways forward (and feel free to simply say which is best):

1. Remove the new codeowners here and hope that we do not have issues on obtaining reviews/approval here in future. This is not ideal from my perspective, but if this is the only acceptable way forward then we can do that for this PR (and if necessary revisit these areas in future).
2. Reduce the new codeowners here to some agreed/limited people.
3. Accept the new codeowners here with the commitment from them to act with appropriate care on these parts of the code.

To be clear, I'm willing to proceed with any of these (or others if suggested).

**Comment by thomasspriggs** [comment]
> I am unhappy with the additions to /util and /goto-programs -- these are very sensitive, and being slow here is good.

@kroening  - There are more than 200 files in `/util` covering a wide variety of different pieces of functionality. Would you be open to a more fine-grained approach to ownership of files within this directory? It could be setup such that certain files are considered to be high risk with a reduced sub-set of code owners and a larger group of code owners for the directory as a whole, for example.

**Comment by kroening** [comment]
> @kroening - There are more than 200 files in `/util` covering a wide variety of different pieces of functionality. Would you be open to a more fine-grained approach to ownership of files within this directory? It could be setup such that certain files are considered to be high risk with a reduced sub-set of code owners and a larger group of code owners for the directory as a whole, for example.

Sure, happy to consider individual files.

**Comment by TGWDB** [comment]
I've pushed an update that removes the contentious additions got `/src/goto-programs` and `/src/util`. For the `/src/util` (at least) I propose this comes in another PR with some exploration/discussion of which files to have special owners for.

**Comment by thomasspriggs** [comment]
I have included some of the SMT2 related changes in my own SMT2 codeowners PR here - https://github.com/diffblue/cbmc/pull/6445

**Comment by TGWDB** [comment]
@kroening Any comment on the updated PR?

**Comment by tautschnig** [comment]
Should we also look at the entries that only have "@diffblue/.*" to ensure the project does not end up being blocked at times where Diffblue cannot commit resources? As a data point, #6367 requires such approval as a new top-level directory was added to the CMake configuration.

**Comment by TGWDB** [comment]
> Should we also look at the entries that only have "@diffblue/.*" to ensure the project does not end up being blocked at times where Diffblue cannot commit resources? As a data point, #6367 requires such approval as a new top-level directory was added to the CMake configuration.

Yes, that's a reasonable suggestion. Did you have any in particular in mind?

**Comment by tautschnig** [comment]
> > Should we also look at the entries that only have "@diffblue/.*" to ensure the project does not end up being blocked at times where Diffblue cannot commit resources? As a data point, #6367 requires such approval as a new top-level directory was added to the CMake configuration.
> 
> Yes, that's a reasonable suggestion. Did you have any in particular in mind?

How about just adding the default code owners?

**Pull request closed without merging**
